src/Orleans/Serialization/SerializationManager.cs

```diff
@@ -863,13 +870,23 @@ public static object DeepCopyInner(object original)
             {
                 copy = copier(original);
                 SerializationContext.Current.RecordObject(original, copy);
+                return copy;
             }
-            else
+
+            var externalSerializer = externalSerializers.FirstOrDefault(serializer => serializer.IsSupportedType(t));
```

Comment:
OK, let me try clarify my ideas:

1. I think the idea of have a global serializer catalog, or a list of them as the initial proposal is something that regardless of other implementation is ok.
2. The serializer registrar should be made at startup time either from a configuration entry inside the silo config and can also discovered by the assembly load step at startup or manually add by code. 
3. Once we have this catalog set up, than there are the ways I propose to use those serializers in order of priority just as OWIN does with its middlewares.

For example:
```
<serializers>
	<serializer type="MyNamespace.MySerializer" assembly="SerializerAssembly"/>
	<serializer type="MyNamespace.MySerializer2" assembly="SerializerAssembly"/>
</serializers>
```

-- Declaratively by attribute on the type regardless if the type is a grain state, a class used to pass to a grain method call, message sent to a stream, whatever.
++++ This will make a specific type to be serialized by the tagged serializer whenever its nevecessary ex. when a message contains this type and it is suppose to be sent cross-silo boundaries or be persisted inside a StorageProvider. 
-- If there is no such attribute on that type, but rather on the grain class, all grain call parameters, state class, or anything that needs be serialized and DON`t have the attribute for a specific serializer, than it will be serialized using the grain one.
++++ This will give a default serializer at grain level if none is provided by the type
-- Define a global fallback serializer by config or code.
++++ This will be the serializer used when all the previous other options didn't match. 
-- If none of the previous match, and there is no global serializer, than just use the first one on the list.

The attribute would be something like [OrleansSerializer(typeof(MySerializer))] Where the serializer class must implement a common serializer interface by the way.

The idea here is to have a level of control over the serialization without need to rely on [Serializable] attribute in order to comply with coreCLR and at the same time keep that simple. The developer could just register the global serializer and everything will behave as is today. If he need one type to be serialized, lets say, encrypted, then just set that specific type with the attribute pointing to SomeEncryptSerializer and so on. This will decouple the serializer once and for all from .net serialization infrastructure and make it available to all other infrastructure components like state, message passing, etc without inject anything. Just get the serializer type from the attribute and get it from the cached serializer type directory and just run it over the type!

So, I guess that is it. Please feedback! :)